src/integrate/scarches/script.py

```diff
@@ -0,0 +1,149 @@
+    logFormatter = logging.Formatter("%(asctime)s %(levelname)-8s %(message)s")
+    console_handler.setFormatter(logFormatter)
+    logger.addHandler(console_handler)
+
+    return logger
+
+PLAN_KWARGS = {
+    "reduce_lr_on_plateau": par['reduce_lr_on_plateau'],
+    "lr_patience": par['lr_patience'],
+    "lr_factor": par['lr_factor'],
+}
+
+
+def map_to_existing_reference(adata_query, model_path, check_val_every_n_epoch=1) -> np.ndarray:
+    """
+    A function to map the query data to the reference atlas
+
+    Input:
+        * adata_query: An AnnData object with the query
+        * model_path: The reference model directory
+
+    Output:
+        A numpy.ndarray containing the embedding coordinates of the query data in the HLCA latent space.
+        If `output_model` is set to True, then the trained reference model is also output
+
+    """
+    print("SCVI-tools version:", scvi.__version__)
+
+    scvi.model.SCANVI.prepare_query_anndata(adata_query, model_path)
+
+    # Load query data into the model
+    vae_query = scvi.model.SCANVI.load_query_data(
+            adata_query,
+            model_path,
+            freeze_dropout=True
+    )
+
+    # Train scArches model for query mapping
+    vae_query.train(
+            max_epochs=par["max_epochs"],
+            early_stopping=par['early_stopping'],
+            early_stopping_monitor=par['early_stopping_monitor'],
+            early_stopping_patience=par['early_stopping_patience'],
+            early_stopping_min_delta=par['early_stopping_min_delta'],
+            check_val_every_n_epoch=check_val_every_n_epoch,
+            use_gpu=(cuda_is_available() or mps_is_available()),
+            plan_kwargs=PLAN_KWARGS
+    )
+
+    return vae_query, adata_query
+
+
+def _download_HLCA_reference_model(directory):
+    import os
+
+    HLCA_PATH = "https://zenodo.org/record/6337966/files/HLCA_reference_model.zip"
+
+    os.system(f"curl  {HLCA_PATH} --output {directory}/HLCA_reference_model.zip")
+    os.system(f"unzip {directory}/HLCA_reference_model.zip -d {directory}")
+
+    scvi.model.SCANVI.convert_legacy_save(f"{directory}/HLCA_reference_model", f"{directory}/HLCA_reference_model_new")
+
+    return f"{directory}/HLCA_reference_model_new"
+
+
+def main():
+    SUPPORTED_REFERENCES = ["HLCA"]
+
+    if par["reference"] is not None and par["reference"] not in SUPPORTED_REFERENCES:
+        raise ValueError(f"{par['reference']} is not supported reference. Please select one of {', '.join(SUPPORTED_REFERENCES)}")
+
+    logger = _setup_logger()
+
+    mdata_query = mudata.read(par["input"].strip())
+    adata_query = mdata_query.mod[par["modality"]]
+
+    if par["reference"] == "HLCA":
+        import tempfile
+
+        with tempfile.TemporaryDirectory() as directory:
+            model_path = _download_HLCA_reference_model(directory)
+            vae_query, adata_query = map_to_existing_reference(adata_query, model_path=model_path)
```

Comment:
this is ideally handled by nextflow... somehow ^^

Reply:
True. One way to do this is to make the `reference` type `file` and set the default to the url (https://zenodo.org/record/6337966/files/HLCA_reference_model.zip). However, this also means that if we want to pass another reference, this also needs to be a path/url. But if we upload them to s3, `s3://` urls are also supported. I think this is a better solution. 

Correct me if im wrong @rcannood

Reply:
I would indeed set the reference to `type: file`. The nextflow module will then https:// and s3:// protocols natively, and will cache the reference model rather than download it during every run.

Reply:
Did it!